### PR TITLE
Add support for returning values from defcal statements

### LINKED ISSUE #21: Add a method to return values from defcal statements
With [#20](https://github.com/openqasm/oqpy/pull/20) merged to main we can now declare return types in defcal statements. However there isn't a convenient way to return values from defcal statements (as far as I can tell, please correct me if I'm wrong). This capability is demonstrated in the [capture instructions](https://openqasm.com/language/openpulse.html#capture-instruction) of the OpenPulse Grammar specification. I suggest implementing a new method of the opqy.Program class that adds a ReturnStatement ast node to the program

```python
def returns(self, return_var:AstConvertible):
    self._add_statement(ast.ReturnStatement(return_var.to_ast(self)))
    return self
```

This method just wraps the current workaround I've found to return a value from a defcal statement which requires using the 'private' _add_statement function of opqy.Program, see e.g. the following example

```python
from openpulse import ast

import oqpy

prog = oqpy.Program()  # create a new oqpy program

m0 = oqpy.PortVar("m0")
cap0 = oqpy.PortVar("cap0")
stimulus_frame = oqpy.FrameVar(m0, 5e9, 0, name="stimulus_frame")
caputure_frame = oqpy.FrameVar(cap0, 5e9, 0, name="capture_frame")
dummy_wfm = oqpy.WaveformVar([-1.0, 0.5, 0.25, -0.75], "dummy_wfm")
capture_v2 = oqpy.declare_extern("capture_v2", [("output", ast.FrameType()), ("duration", oqpy.duration)], oqpy.bit)
qubit = oqpy.PhysicalQubits[0]

with oqpy.defcal(prog, qubit, "measure", return_type=oqpy.bit):
    prog.play(stimulus_frame, dummy_wfm)
    prog._add_statement(
        ast.ReturnStatement(capture_v2(caputure_frame, 16000e-9).to_ast(prog))
    )  # Suggestion: replace with prog.returns(result)

prog.measure(qubit, oqpy.BitVar(name="measured_bit"))

print(prog.to_qasm(encal_declarations=True))
```
Which generates the following OpenQASM code

```
OPENQASM 3.0;
defcalgrammar "openpulse";
cal {
    extern capture_v2(frame, duration) -> bit;
    port m0;
    port cap0;
    frame stimulus_frame = newframe(m0, 5000000000.0, 0);
    waveform dummy_wfm = {-1.0, 0.5, 0.25, -0.75};
    frame capture_frame = newframe(cap0, 5000000000.0, 0);
}
bit measured_bit;
defcal measure $0 -> bit {
    play(stimulus_frame, dummy_wfm);
    return capture_v2(capture_frame, 16000.0ns);
}
measured_bit = measure $0;
```
I'm willing to take the lead on implementing this method and any tests that may be required.

### PR DESCRIPTION
**Pull request by Swordcat**
Added `returns` method to the `oqpy.Program` class, for adding return statements to `defcal` declarations.

The `Program.returns` statement can also be used within `@subroutine` decorated functions instead of the `return` keyword, this opens up the possibility of having branching withing such functions that can have multiple return statements.

All added code covered by tests

Closes #21

### CODE REVIEW

**Comment by Swordcat** [comment]
@PhilReinhold and @jcjaskula-aws would either of you be willing / able to review this PR?

File: oqpy/subroutines.py [hunk 1]
```diff
@@ -99,6 +99,16 @@ def wrapper(program: oqpy.Program, *args: AstConvertible) -> OQFunctionCall:
             body.append(ast.ReturnStatement(to_ast(inner_prog, output)))
         elif output is None:
             return_type = None
+            if type_hints.get("return", False):
+                return_hint = type_hints["return"]()
+                if isinstance(return_hint, _ClassicalVar):
+                    return_type = return_hint
+                elif return_hint is None:
+                    pass
+                else:
```
Comment (PhilReinhold) [inline oqpy/subroutines.py:108]:
```suggestion
                elif return_hint is not None:
```

**Review by PhilReinhold — APPROVED** [review]
Small comment, but otherwise this all looks good to me! I'll give @jcjaskula-aws some time to leave any other comments, but if not I think we can merge.

File: oqpy/subroutines.py [hunk 2]
```diff
@@ -99,6 +99,14 @@ def wrapper(program: oqpy.Program, *args: AstConvertible) -> OQFunctionCall:
             body.append(ast.ReturnStatement(to_ast(inner_prog, output)))
         elif output is None:
             return_type = None
+            if type_hints.get("return", False):
```
Comment (jcjaskula-aws) [inline oqpy/subroutines.py:102]: Please correct me if I'm wrong, this does not treat:
```
def delay50ns(prog: Program, q: Qubit) -> IntVar:
    prog.delay(50e-9, q)
```
or 
```
def delay50ns(prog: Program, q: Qubit):
    prog.delay(50e-9, q)
    prog.returns(1.0)
```
It is not really a blocker. If we don't have a quick workaround, let's open a ticket about these cases.

**Review by jcjaskula-aws — COMMENTED** [review]
Looks very interesting, wrote some small comments. Let's follow or modify the first one on naming, the others can be reported and deferred.

File: oqpy/program.py [hunk 3]
```diff
@@ -405,6 +405,11 @@ def shift_scale(self, frame: AstConvertible, scale: AstConvertible) -> Program:
         self.function_call("shift_scale", [frame, scale])
         return self
 
+    def returns(self, return_var: AstConvertible) -> Program:
```
[the remark this answers is not in the document]
Comment (Swordcat) [reply oqpy/program.py:408]: `return` is a reserved keyword in python https://docs.python.org/3/reference/lexical_analysis.html#keywords, will change `return_var` to `expression` in my next commit

File: oqpy/subroutines.py (hunk 2 above)
Comment (Swordcat) [reply oqpy/subroutines.py:102]: You are correct, I agree we should address them later

File: tests/test_directives.py [hunk 4]
```diff
@@ -661,6 +670,88 @@ def test_defcals():
...
+def test_returns():
+    prog = Program()
+    constant = declare_waveform_generator("constant", [("length", duration), ("iq", complex128)])
+
+    rx_port = PortVar("rx_port")
+    tx_port = PortVar("tx_port")
+    rx_frame = FrameVar(rx_port, 5.752e9, name="rx_frame")
+    tx_frame = FrameVar(tx_port, 5.752e9, name="tx_frame")
+    capture_v2 = oqpy.declare_extern(
+        "capture_v2", [("output", ast.FrameType()), ("duration", oqpy.duration)], oqpy.bit
```
[the remark this answers is not in the document]
Comment (Swordcat) [reply tests/test_directives.py]: I've added port, frame and waveform as aliases for ast.PortType, FrameType and WaveformType

**Review by jcjaskula-aws — APPROVED** [review]

**Pull request merged**